### ISSUE TITLE
How to flow the gradient to each hidden state layer (from 1 to 132 + 1)?

### ISSUE THREAD

**Issue by jasonrichdarmawan** [seal-rg/recurrent-pretraining#27]
My goal is to flow the gradient to each of hidden state layer (e.g., prelude.0 output, prelude.1 output, core_block.0 output of each step, coda.0 output, etc.

~~The problem:~~
~~1. For every 2 core_blocks, the other 2 core_blocks grad is None.~~
    ~~For example, prelude.0, prelude.1, core_block.0 step 0, core_block.1 step 0 have grad. Meanwhile, core_block.2 step 0, core_block.3 step 0 grad is None~~

~~Possible issue:~~
~~1. `self.transformer.ln_f` is called in `iterate_forward` function i.e. it's called twice~~
~~2. `xk.detach()`, which `xk = x`~~

What I've tried:
1. use `x.retain_grad()` within each for loop.

https://github.com/seal-rg/recurrent-pretraining/blob/0d9ed974d253e16498edec5c0c0916fdef4eb339/recpre/raven_modeling_minimal.py#L754C9-L754C39

Edit: Due to GPUs constraints (2x 3090 which is used to store 132 hidden state layers), I use the `device_map={'transformer.wte': 0, ...}`, this breaks the gradient tracking.

This issue was confirmed by:
1. Remove the code from `RavenForCausalLM.forward` and `SandwichBlock.forward`. Frustratingly, the gradient is still missing.
2. Write a dummy code, mimicing `RavelForCausalLM.forward` and `SandwichBlock.forward`. Dear God, I spent 2 days debugging this, always thinking "there is the gradient in dummy code, why it does not work in the actual model?".
3. Then, a simple `"device_map={'transformer.wte': 0, ...}` changed to `device_map='cuda' fixed the issue. Sometimes, I wish I have more compute power so I don't need to touch the `device_map`

Perhaps the fix require modifying the `forward()` function by adding `.to()` function to support multi-GPUs.
